custom_slack_provider/helpers.py

```diff
@@ -0,0 +1,191 @@
+from allauth.socialaccount.models import SocialLogin
+from allauth.socialaccount.providers.base import AuthError, AuthProcess
+
+
+def _process_signup(request, sociallogin):
+    auto_signup = get_adapter(request).is_auto_signup_allowed(
+        request,
+        sociallogin)
+    if not auto_signup:
+        request.session['socialaccount_sociallogin'] = sociallogin.serialize()
+        url = reverse('socialaccount_signup')
+        ret = HttpResponseRedirect(url)
+    else:
+        # Ok, auto signup it is, at least the e-mail address is ok.
+        # We still need to check the username though...
+        if account_settings.USER_MODEL_USERNAME_FIELD:
+            username = user_username(sociallogin.user)
+            try:
+                get_account_adapter(request).clean_username(username)
+            except ValidationError:
+                # This username is no good ...
+                user_username(sociallogin.user, '')
+        # FIXME: This part contains a lot of duplication of logic
+        # ("closed" rendering, create user, send email, in active
+        # etc..)
+        if not get_adapter(request).is_open_for_signup(
+                request,
+                sociallogin):
+            return render(
+                request,
+                "account/signup_closed." +
+                account_settings.TEMPLATE_EXTENSION)
+        get_adapter(request).save_user(request, sociallogin, form=None)
+        ret = complete_social_signup(request, sociallogin)
+    return ret
+
+
+def _login_social_account(request, sociallogin):
+    return perform_login(request, sociallogin.user,
+                         email_verification=app_settings.EMAIL_VERIFICATION,
+                         redirect_url=sociallogin.get_redirect_url(request),
+                         signal_kwargs={"sociallogin": sociallogin})
+
+
+def render_authentication_error(request,
+                                provider_id,
+                                error=AuthError.UNKNOWN,
+                                exception=None,
+                                extra_context=None):
+    try:
+        if extra_context is None:
+            extra_context = {}
+        get_adapter(request).authentication_error(
+            request,
+            provider_id,
+            error=error,
+            exception=exception,
+            extra_context=extra_context)
+    except ImmediateHttpResponse as e:
+        return e.response
+    if error == AuthError.CANCELLED:
+        return HttpResponseRedirect(reverse('socialaccount_login_cancelled'))
+    context = {
+        'auth_error': {
+            'provider': provider_id,
+            'code': error,
+            'exception': exception
+        }
+    }
+    context.update(extra_context)
+    return render(
+        request,
+        "socialaccount/authentication_error." +
+        account_settings.TEMPLATE_EXTENSION,
+        context
+    )
+
+
+def _add_social_account(request, sociallogin):
```

Comment:
how come some functions start with the `_` underscore, but others don't?

Reply:
It is supposed to symbolise a private function. I took it from the actual helpers in the `allauth` module.